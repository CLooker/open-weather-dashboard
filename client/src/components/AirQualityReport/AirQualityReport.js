import React, { PureComponent } from 'react';
import styled from 'styled-components';
import factoryIcon from '../../assets/factory.jpg';
import { apiKey } from '../../utils';
import { IconAndText, Loading } from '../common';
import PollutantData from './PollutantData';

const AirQualityReportWrapper = styled.div`
  border: solid 1px lightgrey;
  border-radius: 10px;
  padding: 15px 0 0;
`;

export default class AirQualityReport extends PureComponent {
  state = {
    co: {
      precision: 0,
      pressure: 0,
      value: 0
    },
    so2: {
      precision: 0,
      pressure: 0,
      value: 0
    }
  };

  componentDidMount() {
    Object.keys(this.state).forEach(pollutant => {
      fetch(
        `https://api.openweathermap.org/pollution/v1/${pollutant}/41,12/current.json?appid=${apiKey}`
      )
        .then(res => res.json())
        .then(res => {
          const { data } = res;
          const { length } = data;

          if (!data || !length) return;

          const accumulatedData = data.reduce(
            (accumulatedData, dataObj) => ({
              precision: accumulatedData.precision + dataObj.precision,
              pressure: accumulatedData.pressure + dataObj.pressure,
              value: accumulatedData.value + dataObj.value
            }),
            this.state[pollutant]
          );
          const averagedData = {
            precision: accumulatedData.precision / length,
            pressure: accumulatedData.pressure / length,
            value: accumulatedData.value / length
          };
          this.setState({ [pollutant]: averagedData });
        })
        .catch(console.err);
    });
  }

  getIsLoading = () =>
    Object.values(this.state).every(pollutant =>
      Object.values(pollutant).every(value => value === 0)
    );

  getAirQualityReportIconAndText = () => {
    const IconAndTextWrapper = styled(IconAndText)`
      display: flex;
      justify-content: center;
      align-items: 'center';
    `;

    return (
      <div>
        <IconAndTextWrapper
          alt='air quality report'
          iconSrc={factoryIcon}
          text={<h3>Air Quality Report</h3>}
        />
      </div>
    );
  };

  getPollutantData = () => {
    const PollutantDataWrapper = styled.div`
      list-style-type: none;
      display: flex;
      justify-content: center;
      align-items: center;
    `;

    return (
      <PollutantDataWrapper>
        {Object.entries(this.state).map(entry => {
          const [pollutant, data] = entry;
          return (
            <PollutantData key={pollutant} pollutant={pollutant} data={data} />
          );
        })}
      </PollutantDataWrapper>
    );
  };

  render() {
    if (this.getIsLoading()) return <Loading />;

    return (
      <AirQualityReportWrapper>
        {this.getAirQualityReportIconAndText()}
        {this.getPollutantData()}
      </AirQualityReportWrapper>
    );
  }
}
