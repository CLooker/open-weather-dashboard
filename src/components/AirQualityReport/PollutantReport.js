import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const PollutantReportWrapper = styled.div`
  margin: 0 5px;
`;

const DataWrapper = styled.div`
  display: flex;
  align-items: center;
  height: 15px;
  margin: 2px 0;
`;

const DataTitleWrapper = styled.h5`
  padding: 0 2px 0 0;
`;

export default class PollutantReport extends PureComponent {
  static propTypes = {
    pollutant: PropTypes.string,
    data: PropTypes.object
  };

  render() {
    const {
      pollutant,
      data: { value, pressure, precision }
    } = this.props;

    return (
      <PollutantReportWrapper>
        <strong title='Pollutant'>{pollutant.toUpperCase()}</strong>
        <DataWrapper>
          <DataTitleWrapper title='Mixing Ratio'>Value:</DataTitleWrapper>
          {value}
        </DataWrapper>
        <DataWrapper>
          <DataTitleWrapper title='Atmospheric Pressure'>
            Pressure:
          </DataTitleWrapper>
          {pressure} hPa
        </DataWrapper>
        <DataWrapper>
          <DataTitleWrapper title='Measurement Precision'>
            Precision:
          </DataTitleWrapper>
          {precision}
        </DataWrapper>
      </PollutantReportWrapper>
    );
  }
}
