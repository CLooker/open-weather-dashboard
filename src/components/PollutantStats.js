import React from 'react';

const PollutantStats = ({ type, chem }) => {
  switch (type) {
    case 'co':
    case 'so2':
      return (
        <div key={type} className="pollutant-stats">
          <div>
            <strong>{type.toUpperCase()}</strong>
          </div>
          <div>Value: {chem.value}</div>
          <div>Pressure: {chem.pressure} hPa </div>
          <div>Precision: {chem.precision}</div>
        </div>
      );
    case 'o3':
      return (
        <div key={type}>
          <div>{type.toUpperCase()}</div>
          <div>Value: {chem.value}</div>
        </div>
      );
    case 'no2':
      return (
        <div key={type}>
          <div>{type.toUpperCase()}</div>
          <div>Value: {chem.value}</div>
          <div>Precision: {chem.precision}</div>
        </div>
      );
    default:
      return null;
  }
};

export default PollutantStats;
