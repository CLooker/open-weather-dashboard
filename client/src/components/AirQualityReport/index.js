import AirQualityReport from './AirQualityReport';

export default AirQualityReport;
