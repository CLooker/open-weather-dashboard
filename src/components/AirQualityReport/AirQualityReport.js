import React, { PureComponent } from 'react';
import styled from 'styled-components';
import { apiKey } from '../../utils';
import { IconAndText, Loading } from '../common';
import PollutantReport from './PollutantReport';

const AirQualityReportWrapper = styled.div`
  border: solid 1px lightgrey;
  border-radius: 10px;
  padding: 15px 0 0;
`;

export default class AirQualityReport extends PureComponent {
  state = {
    co: {
      precision: 0,
      pressure: 0,
      value: 0
    },
    so2: {
      precision: 0,
      pressure: 0,
      value: 0
    }
  };

  componentDidMount() {
    Object.keys(this.state).forEach(pollutant => {
      fetch(
        `https://api.openweathermap.org/pollution/v1/${pollutant}/41,12/current.json?appid=${apiKey}`
      )
        .then(res => res.json())
        .then(res => {
          const { data } = res;
          const { length } = data;

          if (!data || !length) return;

          const accumulatedData = data.reduce(
            (accumulatedData, dataObj) => ({
              precision: accumulatedData.precision + dataObj.precision,
              pressure: accumulatedData.pressure + dataObj.pressure,
              value: accumulatedData.value + dataObj.value
            }),
            this.state[pollutant]
          );
          const averagedData = {
            precision: accumulatedData.precision / length,
            pressure: accumulatedData.pressure / length,
            value: accumulatedData.value / length
          };
          this.setState({ [pollutant]: averagedData });
        })
        .catch(console.err);
    });
  }

  getIsLoading = () =>
    Object.values(this.state).every(pollutant =>
      Object.values(pollutant).every(value => value === 0)
    );

  getAirQualityReportIconAndText = () => {
    const IconAndTextWrapper = styled(IconAndText)`
      display: flex;
      justify-content: center;
      align-items: 'center';
    `;

    return (
      <div>
        <IconAndTextWrapper
          alt='air quality report'
          iconSrc='https://image.freepik.com/icones-gratis/fabrica-de-fumaca-eco_318-41523.jpg'
          text={<h3>Air Quality Report</h3>}
        />
      </div>
    );
  };

  getPollutantReports = () => {
    const PollutantReportsWrapper = styled.div`
      list-style-type: none;
      display: flex;
      justify-content: center;
      align-items: center;
    `;

    return (
      <PollutantReportsWrapper>
        {Object.entries(this.state).map(entry => {
          const [pollutant, data] = entry;
          return (
            <PollutantReport
              key={pollutant}
              pollutant={pollutant}
              data={data}
            />
          );
        })}
      </PollutantReportsWrapper>
    );
  };

  render() {
    if (this.getIsLoading()) return <Loading />;

    return (
      <AirQualityReportWrapper>
        {this.getAirQualityReportIconAndText()}
        {this.getPollutantReports()}
      </AirQualityReportWrapper>
    );
  }
}
