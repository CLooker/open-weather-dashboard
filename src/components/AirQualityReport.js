import React, { Component } from 'react';
import Loading from './Loading';
import PollutantStats from './PollutantStats';
import apiKey from '../api';

export default class AirQualityReport extends Component {
  state = {
    co: {},
    o3: {},
    so2: {},
    no2: {}
  };

  isThereAnyLocalInfo = () =>
    Object.keys(this.state) // state keys
      .map(key => this.state[key]) // state key vals
      .map(obj => Object.keys(obj).length) // coll of state key val lenghts
      .some(l => l > 0); // check if any are non-zero

  componentDidMount() {
    Object.keys(this.state).forEach(key => {
      fetch(
        `http://api.openweathermap.org/pollution/v1/${key}/41,12/current.json?appid=${apiKey}`
      )
        .then(res => res.json())
        .then(
          res =>
            res.message !== 'not found' && this.setState({ [key]: res.data[0] })
        )
        .catch(err => console.log(err));
    });
  }

  render() {
    return (
      <div className="air-quality-report-container">
        {!this.isThereAnyLocalInfo() ? (
          <Loading />
        ) : (
          <div className="air-quality-report">
            <div>
              <img
                src="https://image.freepik.com/icones-gratis/fabrica-de-fumaca-eco_318-41523.jpg"
                alt="pollution"
              />
              <h3>Air Quality Report</h3>
            </div>
            <div className="chem-container">
              {Object.keys(this.state).map(chem => {
                return (
                  Object.keys(this.state[chem]).length > 0 && (
                    <PollutantStats
                      key={chem}
                      type={chem}
                      chem={this.state[chem]}
                    />
                  )
                );
              })}
            </div>
          </div>
        )}
      </div>
    );
  }
}
