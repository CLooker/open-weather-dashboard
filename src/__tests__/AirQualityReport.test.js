import React from 'react';
import { mount, shallow } from 'enzyme';
import AirQualityReport from '../components/AirQualityReport';
import Loading from '../components/Loading';
import PollutantStats from '../components/PollutantStats';

describe('AirQualityReport', () => {
  let component;

  beforeEach(() => (component = shallow(<AirQualityReport />)));

  afterEach(() => jest.restoreAllMocks());

  describe('initialization', () => {
    it('should render correctly', () => {
      expect(component).toMatchSnapshot();
    });

    describe('rendering', () => {
      describe('code paths', () => {
        describe('`isThereAnyLocalInfo` returns `false`', () => {
          it('should render `Loading`', () => {
            component.instance().isThereAnyLocalInfo = jest.fn(() => false);
            const actual = component.html();
            const expected = shallow(<Loading />).html();
            expect(actual).toEqual(expect.stringContaining(expected));
          });
        });

        describe('`isThereAnyLocalInfo` returns true', () => {
          describe('code paths', () => {
            describe('the nested objects in state are empty', () => {
              it('should not render `PollutantStats`', () => {
                const actual = component.find('PollutantStats').length;
                const expected = 0;
                expect(actual).toBe(expected);
              });
            });

            // next
            describe('at least one nested object in state has a value', () => {});
          });
        });
      });
    });
  });
});
