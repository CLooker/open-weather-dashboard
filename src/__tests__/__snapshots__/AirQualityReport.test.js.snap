// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AirQualityReport initialization should render correctly 1`] = `
"<div className=\\"air-quality-report-container\\">
  <Loading />
</div>"
`;
